cli_tools/gce_vm_image_import/importer/importer.go

```diff
@@ -81,46 +92,83 @@ func (i importer) Run(ctx context.Context) (loggable service.Loggable, err error
 }
 
 func (i *importer) runInflate(ctx context.Context) (err error) {
-	select {
-	case <-ctx.Done():
-		err = ctx.Err()
-	default:
+	return i.runStep(ctx, func(ctx context.Context) error {
+		var err error
 		i.pd, err = i.inflater.inflate(ctx)
-		i.traceLogs = append(i.traceLogs, i.inflater.traceLogs()...)
-	}
-	return err
+		return err
+	}, i.inflater.cancel, i.inflater.traceLogs)
 }
 
 func (i *importer) runProcess(ctx context.Context) (err error) {
-	processor, err := i.processorProvider.provide(i.pd)
+	i.processor, err = i.processorProvider.provide(i.pd)
 	if err != nil {
 		return err
 	}
+	return i.runStep(ctx, func(ctx context.Context) error { return i.processor.process(ctx) }, i.processor.cancel, i.processor.traceLogs)
+}
+
+func (i *importer) runStep(ctx context.Context, step func(context.Context) error, cancel func(string), getTraceLogs func() []string) (err error) {
+	e := make(chan error)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		//this select checks if context expired prior to runStep being called
+		//if not, step is run
+		select {
+		case <-ctx.Done():
+			e <- i.getCtxError(ctx)
+		default:
+			stepErr := step(ctx)
+			wg.Done()
+			e <- stepErr
+		}
+	}()
+
+	// this select waits for either context expiration or step to finish (with either an error or success)
 	select {
 	case <-ctx.Done():
-		err = ctx.Err()
-	default:
-		err = processor.process(ctx)
-		i.traceLogs = append(i.traceLogs, processor.traceLogs()...)
+		err = i.getCtxError(ctx)
+		cancel("timed-out")
+		wg.Wait()
+	case inflaterErr := <-e:
+		err = inflaterErr
+	}
+	i.traceLogs = append(i.traceLogs, getTraceLogs()...)
+	return err
+}
+
+func (i *importer) getCtxError(ctx context.Context) (err error) {
+	if ctxErr := ctx.Err(); ctxErr == context.DeadlineExceeded {
+		err = daisy.Errf("Import did not complete within the specified timeout of %s", i.timeout)
+	} else {
+		err = ctxErr
 	}
 	return err
 }
 
 func (i *importer) cleanupDisk() {
-	if i.pd.uri != "" {
-		diskName := path.Base(i.pd.uri)
-		err := i.diskClient.DeleteDisk(i.project, i.zone, diskName)
-		if err != nil {
-			logger.Errorf("Failed to remove temporary disk %v: %e", i.pd, err)
-		}
+	if i.pd.uri == "" {
+		return
+	}
+
+	diskName := path.Base(i.pd.uri)
+
+	disk, err := i.diskClient.GetDisk(i.project, i.zone, diskName)
+	if err != nil || disk == nil {
+		return
+	}
+	err = i.diskClient.DeleteDisk(i.project, i.zone, diskName)
+	if err != nil {
+		logger.Errorf("Failed to remove temporary disk %v: %e", i.pd, err)
 	}
 }
```

Comment:
I took a look, and if the disk doesn't exist, Delete returns a 404. So, I'd recommend doing a single call, and skipping the error message if it's a 404 (and omitting the GetDisk call)

Reply:
Done